### PR TITLE
DAOS-4925 dfuse: Add options to set thread count in dfuse.

### PR DESCRIPTION
**Pull request by ashleypittman**
Add a command line option for the number of threads, or fall back
to any cpuset defined (can be set via numactl or tasket) or one thread
per core on the node if not otherwise constrained.

Signed-off-by: Ashley Pittman <ashley.m.pittman@intel.com>

### CODE REVIEW

**Comment by ashleypittman** [comment]
This needs updating to  check if a local cpuset is in force, or the supplied cpuset is the entire node.

**Comment by ashleypittman** [comment]
I have just tested this with taskset as well, and it also works.  It could use a debug statement to say how many threads were selected, I'll add one in the next iteration.

File: src/client/dfuse/dfuse_main.c [hunk 1]
```diff
@@ -364,6 +371,26 @@ main(int argc, char **argv)
...
+
+		/* Reserve one CPU thread for the daos event queue */
+		dfuse_info->di_thread_count -= 1;
+
+		if (dfuse_info->di_thread_count < 1) {
+			printf("Dfuse needs more threads\n");
+			exit(1);
+		}
+	}
+
```
Comment (ashleypittman) [inline src/client/dfuse/dfuse_main.c:393]: Need to log dfuse_info->di_thread_count here.

File: src/client/dfuse/dfuse_main.c [hunk 2]
```diff
@@ -364,6 +371,26 @@ main(int argc, char **argv)
...
+			exit(1);
+		}
+
+		dfuse_info->di_thread_count = CPU_COUNT(&cpuset);
+
+		/* Reserve one CPU thread for the daos event queue */
+		dfuse_info->di_thread_count -= 1;
+
+		if (dfuse_info->di_thread_count < 1) {
+			printf("Dfuse needs more threads\n");
```
Comment (mchaarawi) [inline src/client/dfuse/dfuse_main.c:389]: print this to stderr?
probably some more info would help (di_thread_count - 1)

**Review by mchaarawi — APPROVED** [review]

File: src/client/dfuse/dfuse_main.c (hunk 2 above)
Comment (ashleypittman) [reply src/client/dfuse/dfuse_main.c:389]: I'd fix the output in a future PR.  I'm adding new command line options and help for login-node support anyway.

File: src/client/dfuse/dfuse_main.c [hunk 3]
```diff
@@ -364,6 +371,26 @@ main(int argc, char **argv)
...
+			printf("Failed to get cpuset information\n");
+			exit(1);
+		}
+
+		dfuse_info->di_thread_count = CPU_COUNT(&cpuset);
+
+		/* Reserve one CPU thread for the daos event queue */
+		dfuse_info->di_thread_count -= 1;
+
+		if (dfuse_info->di_thread_count < 1) {
```
Comment (daltonbohning) [inline src/client/dfuse/dfuse_main.c:388]: I think we need a similar check for the passed in `--thread-count`. Right now, dfuse lets me pass in 0 threads.
Maybe check this outside of this if statement to account for the passed in value or cpuset?

**Review by daltonbohning — COMMENTED** [review]
`c = getopt_long(argc, argv, "s:m:Sfh"`
Needs `t` to accept short option at the command line

**Review by jolivier23 — APPROVED** [review]

File: src/client/dfuse/dfuse_main.c (hunk 3 above)
Comment (ashleypittman) [reply src/client/dfuse/dfuse_main.c:388]: I'll fix this in https://github.com/daos-stack/daos/pull/3894 which is changing the command line processing anyway.

**Comment by ashleypittman** [comment]
Test failures are DAOS-6022 and DAOS-6074

**Review by daltonbohning — APPROVED** [review]

**Pull request merged**